### ISSUE TITLE
Tie POI names to icons

### ISSUE THREAD

**Issue by pnorman** [gravitystorm/openstreetmap-carto#234]
If a POI icon conflicts with another label but not its text right now the text will be rendered without any icon. Examples have shown up in #232 and maybe #201 but I've seen it locally.

A simple example is
![18/41567/89792](https://f.cloud.github.com/assets/1190866/1349524/649dc1a8-36f0-11e3-8760-33f972f37030.png)
The restaurant text conflicts with the icon. Also, the text is not centered on the POI because it's offset to below to allow for an icon.

A more complex example is 
![18/41565/89790](https://f.cloud.github.com/assets/1190866/1349507/e3096a0c-36ef-11e3-9e0b-747065c60916.png)

[AFG Kabul Bakery](http://www.openstreetmap.org/browse/node/960913456) (`shop=bakery`) has its label rendered, but the icon conflicts with the icon for [Zahra's Ultimate Hair Design](http://www.openstreetmap.org/browse/node/960913382) (`shop = hairdresser`). With the name it's fairly obvious what AFG Kabul Bakery is, but there's no icon to it. 

Oddly, it looks like the text for Zahra's Ultimate Hair Design conflicts with the AFG Kabul Bakery text.

It would be an improvement if the POI name was only rendered in conjunction with the icon, and if there wasn't room for the icon it didn't render either.

This will probably not impact rendered POI density significantly because in dense areas where this is an issue the blank space from removing the name will probably be taken up by another icon.

**Comment by daganzdaanda** [comment]
> I honestly cannot recall this many detached names in the old style. 
Yes, I get the feeling that this has increased recently in my home area. And I don't think that there were many shops added. Right now, doctors begin to show up in the tiles, which is great. But this adds many more icons.

Maybe the distance between icon and label (text-dy?) has been increased too much? Especially the generic shop dot is too far away with most Roman labels. Since we can't use a dynamic distance that follows the actual letters (or can we?), hopefully #957 will be possible somehow. It may reduce these "orphan" icons significantly.

**Comment by boothym** [comment]
Noticed this happening to bus stops and memorials with names missing, example here: https://www.openstreetmap.org/node/502530719#map=17/55.95858/-3.20779 

@nebulon42 hoping your PR will solve it :)

**Comment by nebulon42** [comment]
Unfortunately, that's unlikely. My PR would solve labels without icons. At the link it looks like the bus stop label is blocked by the address.

**Comment by nebulon42** [comment]
> Am I mistaken or isn't that exactly what my link is, a bus stop label without an icon?

Hm, ok, I'm only seeing a bus icon without label on z18 but never mind. The memorial I see too. Yes, that should be solved by #2597.

**Comment by matthijsmelissen** [comment]
We currently have no good way to accomplish this, so I am going to close this issue.

> At that point though, I wonder whether it's better to introduce a new feature in mapnik to have a grouped-symbolizer with controllable characteristics.

I agree. @springmeyer what do you think of this? I can imagine others struggle with the same problem so I think this would be nice to have in Mapnik.

**Comment by dieterdreist** [comment]
this is already available for a very long time: https://github.com/mapnik/mapnik/wiki/ShieldSymbolizer

sent from a phone

**Comment by matthijsmelissen** [comment]
No, that is not what we need: shield symbolizers prevent rendering text without icon, however they also prevent rendering icons without text. We don't want the latter.

**Comment by nebulon42** [comment]
I did already show that it is possible to have both in #2597. However, there is also a GroupSymbolizer in Mapnik 3, but someone would have to implement it for CartoCSS.

**Comment by dieterdreist** [comment]
sent from a phone

> On 8. May 2018, at 03:01, Matthijs Melissen <notifications@github.com> wrote:
> 
> No, that is not what we need: shield symbolizers prevent rendering text without icon, however they also prevent rendering icons without text. We don't want the latter.


you could first try the shields and after that try just markers to fill up the gaps. Admittedly there is a performance penalty involved.

**Comment by pnorman** [comment]
> you could first try the shields and after that try just markers to fill up the gaps. Admittedly there is a performance penalty involved.

Is this penalty significant?

**Comment by jeisenbe** [comment]
#2597 was interesting. I've noticed this problem is particularly acute with aerodromes (airports) at low zoom levels in my area. The airport is often right next to the town, so the icon may be blocked by the name of the town, but the purple text label renders right below the town name.

**Comment by kocio-pl** [comment]
From https://github.com/gravitystorm/openstreetmap-carto/issues/234#issuecomment-388033294:
> However, there is also a GroupSymbolizer in Mapnik 3, but someone would have to implement it for CartoCSS.

I understand that would make icons with optional labels possible? Is there anybody willing to do that?
